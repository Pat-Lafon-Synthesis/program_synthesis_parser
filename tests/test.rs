use std::{fs::File, io::Read};

use program_synthesis_parser;

macro_rules! make_test {
    ($test_name:tt) => {
        #[test]
        fn $test_name() {
            let parser = program_synthesis_parser::spec::ProblemParser::new();
            let mut file =
                File::open(format!("tests/benchmarks/{}.mls", stringify!($test_name))).unwrap();
            let mut buffer = String::new();
            file.read_to_string(&mut buffer).unwrap();
            insta::assert_debug_snapshot!(parser.parse(&buffer).unwrap());
        }
    };
}

make_test!(bool_always_false);
make_test!(bool_always_true);
make_test!(bool_band);
make_test!(bool_bor);
make_test!(bool_impl);
make_test!(bool_neg);


make_test!(expr_boolean);
make_test!(expr_div);
make_test!(expr_sub);
make_test!(expr);


make_test!(list_append);
make_test!(list_compress);
make_test!(list_concat);
make_test!(list_drop);
make_test!(list_dropeven);
make_test!(list_even_parity);
make_test!(list_filter);
make_test!(list_fold);
make_test!(list_hd);
make_test!(list_inc);
make_test!(list_last);
make_test!(list_last2);
make_test!(list_length);
make_test!(list_make);
make_test!(list_map);
make_test!(list_nth);
make_test!(list_pairwise_swap);
make_test!(list_range);
make_test!(list_rev_append);
make_test!(list_rev_fold);
make_test!(list_rev_snoc);
make_test!(list_rev_tailcall);
make_test!(list_snoc);
make_test!(list_sort_sorted_insert);
make_test!(list_sorted_insert);
make_test!(list_stutter);
make_test!(list_sum);
make_test!(list_take);
make_test!(list_tl);


make_test!(nat_add);
make_test!(nat_iseven);
make_test!(nat_max);
make_test!(nat_mul);
make_test!(nat_pred);
make_test!(nat_sub);


make_test!(tree_balanced);
make_test!(tree_binsert);
make_test!(tree_collect_leaves);
make_test!(tree_count_leaves);
make_test!(tree_count_nodes);
make_test!(tree_height);
make_test!(tree_inorder);
make_test!(tree_lastleft);
make_test!(tree_map);
make_test!(tree_nodes_at_level);
make_test!(tree_notexist);
make_test!(tree_postorder);
make_test!(tree_preorder);
make_test!(tree_sum);