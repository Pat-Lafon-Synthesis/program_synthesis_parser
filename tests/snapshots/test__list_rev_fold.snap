---
source: tests/test.rs
expression: parser.parse(&buffer).unwrap()
---
SynthProblem {
    imports: [],
    decls: [
        TypeDecl(
            "nat",
            Type(
                Variant([("O", Type(Tuple([]))), ("S", Type(Named("nat")))]),
            ),
        ),
        TypeDecl(
            "list",
            Type(
                Variant([("Nil", Type(Tuple([]))), ("Cons", Type(Tuple([Type(Named("nat")), Type(Named("list"))])))]),
            ),
        ),
        ExpDecl(
            "snoc",
            Exp(
                Fix("snoc", Type(Arrow(Type(Arrow(Type(Named("list")), Type(Named("nat")))), Type(Named("list")))), Exp(Func(Param { name: "l1", ty: Type(Named("list")) }, Exp(Func(Param { name: "n", ty: Type(Named("nat")) }, Exp(Match(Exp(Var("l1")), [(Ctor("Nil", Wildcard), Exp(CTor("Cons", Exp(Tuple([Exp(Var("n")), Exp(CTor("Nil", Exp(Tuple([]))))]))))), (Ctor("Cons", Tuple([Var("h"), Var("t")])), Exp(CTor("Cons", Exp(Tuple([Exp(Var("h")), Exp(App(Exp(App(Exp(Var("snoc")), Exp(Var("t")))), Exp(Var("n"))))])))))]))))))),
            ),
        ),
        ExpDecl(
            "fold",
            Exp(
                Fix("fold", Type(Arrow(Type(Arrow(Type(Arrow(Type(Arrow(Type(Arrow(Type(Named("list")), Type(Named("nat")))), Type(Named("list")))), Type(Named("list")))), Type(Named("list")))), Type(Named("list")))), Exp(Func(Param { name: "f", ty: Type(Arrow(Type(Arrow(Type(Named("list")), Type(Named("nat")))), Type(Named("list")))) }, Exp(Func(Param { name: "init", ty: Type(Named("list")) }, Exp(Func(Param { name: "l", ty: Type(Named("list")) }, Exp(Match(Exp(Var("l")), [(Ctor("Nil", Wildcard), Exp(Var("init"))), (Ctor("Cons", Tuple([Var("h"), Var("t")])), Exp(App(Exp(App(Exp(Var("f")), Exp(App(Exp(App(Exp(App(Exp(Var("fold")), Exp(Var("f")))), Exp(Var("init")))), Exp(Var("t")))))), Exp(Var("h")))))]))))))))),
            ),
        ),
    ],
    synth_type: Type(
        Arrow(Type(Named("list")), Type(Named("list"))),
    ),
    spec: UIOEs(
        [
            Example {
                input: [
                    Exp(
                        CTor("Nil", Exp(Tuple([]))),
                    ),
                ],
                output: Exp(
                    CTor("Nil", Exp(Tuple([]))),
                ),
            },
            Example {
                input: [
                    Exp(
                        CTor("Cons", Exp(Tuple([Exp(CTor("O", Exp(Tuple([])))), Exp(CTor("Cons", Exp(Tuple([Exp(CTor("S", Exp(CTor("O", Exp(Tuple([])))))), Exp(CTor("Nil", Exp(Tuple([]))))]))))]))),
                    ),
                ],
                output: Exp(
                    CTor("Cons", Exp(Tuple([Exp(CTor("S", Exp(CTor("O", Exp(Tuple([])))))), Exp(CTor("Cons", Exp(Tuple([Exp(CTor("O", Exp(Tuple([])))), Exp(CTor("Nil", Exp(Tuple([]))))]))))]))),
                ),
            },
        ],
    ),
}
